plugins/commands/upload/command.rb

```diff
@@ -0,0 +1,231 @@
+require 'optparse'
+require "rubygems/package"
+
+module VagrantPlugins
+  module CommandUpload
+    class Command < Vagrant.plugin("2", :command)
+
+      VALID_COMPRESS_TYPES = [:tgz, :zip].freeze
+
+      def self.synopsis
+        "upload to machine via communicator"
+      end
+
+      def execute
+        options = {}
+
+        opts = OptionParser.new do |o|
+          o.banner = "Usage: vagrant upload [options] <source> [destination] [name|id]"
+          o.separator ""
+          o.separator "Options:"
+          o.separator ""
+
+          o.on("-t", "--temporary", "Upload source to temporary directory") do |t|
+            options[:temporary] = t
+          end
+
+          o.on("-c", "--compress", "Use gzip compression for upload") do |c|
+            options[:compress] = c
+          end
+
+          o.on("-C", "--compression-type=TYPE", "Type of compression to use (#{VALID_COMPRESS_TYPES.join(", ")})") do |c|
+            options[:compression_type] = c.to_sym
+            options[:compress] = true
+          end
+        end
+
+        argv = parse_options(opts)
+        return if !argv
+
+        case argv.size
+        when 3
+          source, destination, guest = argv
+        when 2, 1
+          source = argv[0]
+          if @env.active_machines.map(&:first).include?(argv[1])
+            guest = argv[1]
+          else
+            destination = argv[1]
+          end
+          destination ||= File.basename(source)
+        else
+          raise Vagrant::Errors::CLIInvalidUsage, help: opts.help.chomp
+        end
```

Comment:
Hmmm, I think this case statement might need to be reworked. If you want to upload something to the default path and you're in a multi-vm environment, you'll need to specify the path no matter what:

```
brian@localghost:vagrant-sandbox % be vagrant upload installers bork                                     ±[●][master]
This command requires a specific VM name to target in a multi-VM environment.
```

Maybe it makes sense to move the vm argument to the beginning? I think other commands do that....maybe :thinking: